File: day7/part_2_solution.rb
def read_file(file_name)
  file = File.open(file_name, "r")
  data = file.read
  file.close
  return data
end


file_name = "input.txt"
arr = read_file(file_name).split(",").map(&:to_i)
min_fuel = 999999999999999
for j in arr.min..arr.max
  fuel = 0
  for i in 0...arr.size
    r = [arr[i], j].max - [arr[i], j].min
    extra = (r * (r+1))/2
    fuel += extra
  end
  min_fuel = [min_fuel, fuel].min
end

p min_fuel

File: 2022/day7/part_2_solution.js
const fs = require('fs');

const readFile = (path) => {
  return fs.readFileSync(path, { encoding: 'utf8' });
}

const solution = (input) => {
  let arr = input.split("\n");
  const max = 100000;
  let sum = 0;
  let dirList = [];
  const dirSize = {};

  for(let i=0; i<arr.length; i++) {
    let line = arr[i];
    if(line.startsWith("$ cd ..")) {
      dirList.pop();
    } else if(line.startsWith("$ cd")) {
      const dir = line.split("$ cd ")[1].trim();
      dirList.push(dir);
    }
    if(line.startsWith("$ ls")) {
      i++;
      do {
        let line = arr[i];
        i++;
        if(line.startsWith("dir")) {
        } else {
          const val = parseInt(line.split(" ")[0]);
          let dirName = "";
          dirList.forEach((dir,i) => {
            dirName += dir + "/";
            if(!dirSize[dirName]) {
              dirSize[dirName] = 0;
            }
            dirSize[dirName] += val;
          });
        }
      } while(i < arr.length && !arr[i].startsWith("$"));
      i--;
    }
  }
  let min = 70000000;
  const req = dirSize['//'] - 40000000;
  Object.keys(dirSize).forEach((key) => {
    if(dirSize[key] >= req) {
      min = Math.min(min, dirSize[key]);
    }
  })

  console.log(min);
}

let fileName = "example";

if (process.argv[2]) {
  const maybe = process.argv[2].split("--input=");
  if(maybe.length === 2) {
    fileName = maybe[1];
  }
}

const fileContent = readFile(`./${fileName}.txt`);
solution(fileContent);